okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java

```diff
@@ -0,0 +1,177 @@
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.netflix.concurrency.limits.Limiter;
+import com.netflix.concurrency.limits.limit.AIMDLimit;
+import com.netflix.concurrency.limits.limit.TracingLimitDecorator;
+import com.netflix.concurrency.limits.limiter.DefaultLimiter;
+import com.netflix.concurrency.limits.strategy.SimpleStrategy;
+import com.palantir.remoting3.tracing.okhttp3.OkhttpTraceInterceptor;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import okhttp3.Request;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Remoting calls may observe 429 or 503 responses from the server, at which point they back off in order to
+ * reduce excess load. Unfortunately this state on backing off is stored per-call, so 429s or 503s in one call do not
+ * cause any request rate slowdown in subsequent calls. This class affects this by adjusting the number of requests
+ * that might be dispatched to a given endpoint.
+ * <p>
+ * This is based on Netflix's <a href="https://github.com/Netflix/concurrency-limits/">Concurrency Limits</a> library,
+ * which provides a number of primitives for this.
+ * <p>
+ * In order to use this class, one should get a Limiter for their request, which returns a future. once the Future is
+ * done, the caller can assume that the request is schedulable. After the request completes, the caller <b>must</b>
+ * call one of the methods on {@link Limiter.Listener} in order to provide feedback about the request's success.
+ * If this is not done, throughput will be negatively affected. We attempt to eventually recover to avoid a total
+ * deadlock, but this is not guaranteed.
+ */
+final class ConcurrencyLimiters {
+    private static final Logger log = LoggerFactory.getLogger(ConcurrencyLimiters.class);
+    private static final String FALLBACK = "";
+
+    // If a request is never marked as complete and is thrown away, recover on the next GC instead of deadlocking
+    private static final Map<Limiter.Listener, Runnable> activeListeners = CacheBuilder.newBuilder()
+            .weakKeys()
+            .<Limiter.Listener, Runnable>removalListener(notification -> {
+                if (notification.getCause().equals(RemovalCause.COLLECTED)) {
+                    log.warn("Concurrency limiter was leaked."
+                            + " This implies a remoting bug or classpath issue, and may cause degraded performance");
+                    notification.getValue().run();
+                }
+            })
+            .build()
+            .asMap();
+
+    private final ConcurrentMap<String, ConcurrencyLimiter> limiters = new ConcurrentHashMap<>();
+
+    private static Limiter<Void> newLimiter() {
```

Comment:
inline in ConcurrencyLimiter constructor?

Reply:
done